### DISCUSSION TITLE
Deno support

### DISCUSSION THREAD

**Discussion by brundonsmith** [mobxjs/mobx#3129] [Ideas]
I'm interested in using MobX in Deno. I know this isn't exactly MobX's primary use-case, but I *think* it should be fairly straightforward to support it since MobX doesn't have any dependencies.

For those who don't know, Deno can ingest ESModules/TypeScript directly. It can also import directly from URLs. So a common practice is to import directly from github:
```
import { observable } from "https://raw.githubusercontent.com/mobxjs/mobx/main/packages/mobx/src/mobx.ts";
```

This got as far as trying to import the other modules (utils, errors, etc) before failing. It failed at those because:
```
error: Import 'https://raw.githubusercontent.com/mobxjs/mobx/main/packages/mobx/src/utils/global' failed: 404 Not Found
    at https://raw.githubusercontent.com/mobxjs/mobx/main/packages/mobx/src/mobx.ts:19:0
```
GitHub URLs require file extensions, and it happens that Deno does too. This *could* be the only problem, or it's possible others will turn up if MobX is internally using any bespoke TypeScript options or fancy module-resolution strategies. However, if this *is* the only problem, it could be as simple as adding file extensions to all internal imports and tweaking the bundler config if necessary to accommodate this. I'd tend to think this wouldn't be a breaking change since most people use a bundled version of MobX, and it would also help get MobX ready for an ESModule future where (I think) those file extensions are also required.

I'd be happy to do legwork on this if there's potential interest in merging such changes! I'm also happy to listen to alternate suggestions on how MobX could be made Deno-ready

**Comment by brundonsmith** [comment]
Update: Just to see, I forked MobX and changed all the imports. There are definitely some other issues, for example:
```typescript
TS7006 [ERROR]: Parameter 'thing' implicitly has an 'any' type.
export function onBecomeObserved(thing, arg2, arg3?): Lambda {
                                 ~~~~~
at https://raw.githubusercontent.com/brundonsmith/mobx/main/packages/mobx/src/api/become-observed.ts:31:34  
    
    
TS2304 [ERROR]: Cannot find name '__DEV__'.
if (__DEV__ && globalState.disableErrorBoundaries === true) {
    ~~~~~~~
at https://raw.githubusercontent.com/brundonsmith/mobx/main/packages/mobx/src/api/configure.ts:53:9


TS1205 [ERROR]: Re-exporting a type when the '--isolatedModules' flag is provided requires using 'export type'.     
IMapDidChange,
~~~~~~~~~~~~~
at https://raw.githubusercontent.com/brundonsmith/mobx/main/packages/mobx/src/mobx.ts:65:5
```

Each of these errors has dozens of instances, and there are other types of errors not listed here. So looks like this would not be a simple task, sadly :( I'd still be excited and willing to help if the project ever decided to address this use-case, but as it stands I understand if this isn't a good use of time.

**Comment by mweststrate** [comment]
That merely looks like the tsconfig that is applied in your fork is
different than that is applied to the original. This might be an issue with
how you compile or bundle things? Does deno automatically pick up tsconfig
of the project, or is there one 'master' setting everything needs to apply
to?

**Comment by brundonsmith** [comment]
You know what, Deno does have a way to override certain compilerOptions. Let me try doing that when I have some time

**Comment by brundonsmith** [comment]
I wasn't able to get this working easily, however, I was able to use MobX in Deno by importing a bundled version from JSPM like so:
```typescript
import { observable, autorun, configure } from "https://jspm.dev/mobx"
```

which I think will work for my purposes, even though it lacks types right now
